File: OneSignalSDK/onesignal/src/main/java/com/onesignal/NotificationPayloadProcessorHMS.java
package com.onesignal;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import static com.onesignal.GenerateNotification.BUNDLE_KEY_ACTION_ID;

class NotificationPayloadProcessorHMS {

    static void handleHMSNotificationOpenIntent(@NonNull Activity activity, @Nullable Intent intent) {
        OneSignal.initWithContext(activity.getApplicationContext());
        if (intent == null)
            return;

        JSONObject jsonData = covertHMSOpenIntentToJson(intent);
        if (jsonData == null)
            return;

        handleProcessJsonOpenData(activity, jsonData);
    }

    // Takes in a Notification Open Intent fired from HMS Core and coverts it to an OS formatted JSONObject
    // Returns null if it is NOT a notification sent from OneSignal's backend
    private static @Nullable JSONObject covertHMSOpenIntentToJson(@Nullable Intent intent) {
        // Validate Intent to prevent any side effects or crashes
        //    if triggered outside of OneSignal for any reason.
        if (!OSNotificationFormatHelper.isOneSignalIntent(intent))
            return null;

        Bundle bundle = intent.getExtras();
        JSONObject jsonData = NotificationBundleProcessor.bundleAsJSONObject(bundle);
        reformatButtonClickAction(jsonData);

        return jsonData;
    }

    // Un-nests JSON, key actionId, if it exists under custom
    // Example:
    //   From this:
    //      { custom: { actionId: "exampleId" } }
    //   To this:
    //      { custom: { }, actionId: "exampleId" } }
    private static void reformatButtonClickAction(@NonNull JSONObject jsonData) {
        try {
            JSONObject custom = NotificationBundleProcessor.getCustomJSONObject(jsonData);
            String actionId = (String)custom.remove(BUNDLE_KEY_ACTION_ID);
            if (actionId == null)
                return;

            jsonData.put(BUNDLE_KEY_ACTION_ID, actionId);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private static void handleProcessJsonOpenData(@NonNull Activity activity, @NonNull JSONObject jsonData) {
        if (NotificationOpenedProcessor.handleIAMPreviewOpen(activity, jsonData))
            return;

        OneSignal.handleNotificationOpen(
            activity,
            new JSONArray().put(jsonData),
            true,
            OSNotificationFormatHelper.getOSNotificationIdFromJson(jsonData)
        );
    }

    // HMS notification with Message Type being Message won't trigger Activity reverse trampolining logic
    // for this case OneSignal rely on NotificationOpenedActivityHMS activity
    // Last EMUI (12 to the date) is based on Android 10, so no
    // Activity trampolining restriction exist for HMS devices
    public static void processDataMessageReceived(@NonNull final Context context, @Nullable String data) {
        OneSignal.initWithContext(context);
        if (data == null)
            return;

        final Bundle bundle = OSUtils.jsonStringToBundle(data);
        if (bundle == null)
            return;

        NotificationBundleProcessor.ProcessBundleReceiverCallback bundleReceiverCallback = new NotificationBundleProcessor.ProcessBundleReceiverCallback() {

            @Override
            public void onBundleProcessed(@Nullable NotificationBundleProcessor.ProcessedBundleResult processedResult) {
                // TODO: Figure out the correct replacement or usage of completeWakefulIntent method
                //      FCMBroadcastReceiver.completeWakefulIntent(intent);

                // Return if the notification will NOT be handled by normal GcmIntentService display flow.
                if (processedResult != null && processedResult.processed())
                    return;

                // TODO: 4.0.0 or after - What is in GcmBroadcastReceiver should be split into a shared class to support FCM, HMS, and ADM
                FCMBroadcastReceiver.startFCMService(context, bundle);
            }
        };
        NotificationBundleProcessor.processBundleFromReceiver(context, bundle, bundleReceiverCallback);

    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/onesignal/internal/notification/work/NotificationPayloadProcessorHMS.kt
package com.onesignal.onesignal.internal.notification.work

import com.onesignal.OSInAppMessagePreviewHandler.notificationOpened
import android.app.Activity
import android.content.Context
import android.content.Intent
import com.onesignal.onesignal.OneSignal
import com.onesignal.onesignal.internal.common.JSONUtils
import com.onesignal.onesignal.internal.notification.NotificationConstants
import com.onesignal.onesignal.internal.notification.NotificationFormatHelper
import com.onesignal.onesignal.internal.notification.NotificationHelper
import org.json.JSONException
import org.json.JSONObject

internal object NotificationPayloadProcessorHMS {
    fun handleHMSNotificationOpenIntent(activity: Activity, intent: Intent?) {
        OneSignal.initWithContext(activity.applicationContext)
        if (intent == null) return
        val jsonData = covertHMSOpenIntentToJson(
            intent
        )
            ?: return
        handleProcessJsonOpenData(activity, jsonData)
    }

    // Takes in a Notification Open Intent fired from HMS Core and coverts it to an OS formatted JSONObject
    // Returns null if it is NOT a notification sent from OneSignal's backend
    private fun covertHMSOpenIntentToJson(intent: Intent?): JSONObject? {
        // Validate Intent to prevent any side effects or crashes
        //    if triggered outside of OneSignal for any reason.
        if (!NotificationFormatHelper.isOneSignalIntent(intent)) return null
        val bundle = intent!!.extras
        val jsonData = JSONUtils.bundleAsJSONObject(bundle!!)
        reformatButtonClickAction(jsonData)
        return jsonData
    }

    // Un-nests JSON, key actionId, if it exists under custom
    // Example:
    //   From this:
    //      { custom: { actionId: "exampleId" } }
    //   To this:
    //      { custom: { }, actionId: "exampleId" } }
    private fun reformatButtonClickAction(jsonData: JSONObject) {
        try {
            val custom = NotificationHelper.getCustomJSONObject(jsonData)
            val actionId = custom.remove(NotificationConstants.GENERATE_NOTIFICATION_BUNDLE_KEY_ACTION_ID) as String
                ?: return
            jsonData.put(NotificationConstants.GENERATE_NOTIFICATION_BUNDLE_KEY_ACTION_ID, actionId)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    private fun handleProcessJsonOpenData(activity: Activity, jsonData: JSONObject) {
        if (notificationOpened(activity, jsonData))
            return

        //TODO: Implement
//        OneSignal.handleNotificationOpen(
//            activity,
//            JSONArray().put(jsonData),
//            NotificationFormatHelper.getOSNotificationIdFromJson(jsonData)
//        )
    }

    // HMS notification with Message Type being Message won't trigger Activity reverse trampolining logic
    // for this case OneSignal rely on NotificationOpenedActivityHMS activity
    // Last EMUI (12 to the date) is based on Android 10, so no
    // Activity trampolining restriction exist for HMS devices
    fun processDataMessageReceived(context: Context, data: String?) {
        if (data == null)
            return

        val bundle = JSONUtils.jsonStringToBundle(data) ?: return

        val bundleProcessor = OneSignal.getService<INotificationBundleProcessor>()

        bundleProcessor.processBundleFromReceiver(context, bundle,)
    }
}